filters/oscillatory_hallucination/filter.py

```diff
@@ -0,0 +1,243 @@
+            or (cp >= 0x2F800 and cp <= 0x2FA1F)  #
+        ):  #
+            return True
+
+        return False
+
+    def _is_control(self, char):
+        """Checks whether `chars` is a control character."""
+        # These are technically control characters but we count them as whitespace
+        # characters.
+        if char == "\t" or char == "\n" or char == "\r":
+            return False
+        cat = unicodedata.category(char)
+        if cat.startswith("C"):
+            return True
+        return False
+
+    def _is_punctuation(self, char):
+        """Checks whether `chars` is a punctuation character."""
+        cp = ord(char)
+        # We treat all non-letter/number ASCII as punctuation.
+        # Characters such as "^", "$", and "`" are not in the Unicode
+        # Punctuation class but we treat them as punctuation anyways, for
+        # consistency.
+        if (
+            (cp >= 33 and cp <= 47)
+            or (cp >= 58 and cp <= 64)
+            or (cp >= 91 and cp <= 96)
+            or (cp >= 123 and cp <= 126)
+        ):
+            return True
+        cat = unicodedata.category(char)
+        if cat.startswith("P"):
+            return True
+        return False
+
+    def _is_whitespace(self, char):
+        """Checks whether `chars` is a whitespace character."""
+        # \t, \n, and \r are technically contorl characters but we treat them
+        # as whitespace since they are generally considered as such.
+        if char == " " or char == "\t" or char == "\n" or char == "\r":
+            return True
+        cat = unicodedata.category(char)
+        if cat == "Zs":
+            return True
+        return False
+
+    def _clean_text(self, text):
+        """Performs invalid character removal and whitespace cleanup on text."""
+        output = []
+        for char in text:
+            cp = ord(char)
+            if cp == 0 or cp == 0xFFFD or self._is_control(char):
+                continue
+            if self._is_whitespace(char):
+                output.append(" ")
+            else:
+                output.append(char)
+        return "".join(output)
+
+
+class OscillatoryHallucinationFilter(SentenceAndTargetOperation):
+
+    """N-gram Count based Heuristic for Detecting Oscillatory Hallucinations
+    Paper: https://arxiv.org/pdf/2104.06683.pdf
+    The paper did not explicitly tokenize since IWSLT is available in tokenized form.
+    Tokenization used here = Whitespace + Punctuation, as in multilinbgual BERT pre-tokenization.
+    Finally, 1 more difference with the paper: Count Threshold is also set along with Difference Threshold.
+    Thresholds are set to very High values to ensure very High Precision of the Filter, but this can be tweaked.
+    """
+
+    tasks = [TaskType.TEXT_TO_TEXT_GENERATION]
+
+    def __init__(self):
+        super().__init__()
+        self.tokenizer = BasicTokenizer()
+        self.ngram_size = 2
+        self.count_threshold = 10
```

Comment:
Should ngram_size, count_threshold and difference_threshold be arguments in init so the users can tweak the thresholds?

Reply:
> Should `ngram_size`, `count_threshold `and `difference_threshold `be arguments in init so the users can tweak the thresholds?

I agree with @timothy22000, these parameters should be configurable.

Reply:
Thanks @timothy22000 , @uyaseen , I have made the above parameters configurable now.